### PR TITLE
ADR to merge the repos

### PR DESCRIPTION
**Pull request by HaGuesto**
I talked about this already a few times and finally had a bit of time to create this adr. Looking forward to your reviews.

### CODE REVIEW

File: docs/adr/adr_monorepo.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+- deployment is easier since only one script is needed.
+- management of frontend vs. backend versioning is reduced to a minimum.
+
+*Con:*
+Technically, there are problems that come with a Mono-repo set up: slow code search, time for testing, chaotic repo,...
+However, we argue that all the problems really surface with a much larger codebase and team.
+
+## Consequences
+
+James is working on a step-by-step plan to move the two repos together.
```
Comment (jamescrowley) [inline docs/adr/adr_monorepo.md:46]: James *will* work on a plan ^^, but obviously only if you folks decide that this is the direction you want to go  :grin:

File: docs/adr/adr_monorepo.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+- Mono-Repo
+- Poly-Repo (current state)
+
+## Decision
+
+Mono-Repo
+
+*Pro:*
+- The main reason to move to a monorepo set-up is to force us to communicate/coordinate with each other and to feel a shared responsibility for the monorepo. :)
```
Comment (jamescrowley) [inline docs/adr/adr_monorepo.md:34]: should be able to more easily run this all with a single docker-compose command too, I think? not sure if that works right now or not?

File: docs/adr/adr_monorepo.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+other side-effects
+- deployment is easier since only one script is needed.
+- management of frontend vs. backend versioning is reduced to a minimum.
+
+*Con:*
+Technically, there are problems that come with a Mono-repo set up: slow code search, time for testing, chaotic repo,...
+However, we argue that all the problems really surface with a much larger codebase and team.
```
Comment (aerinsol) [inline docs/adr/adr_monorepo.md:42]: Be specific with what you mean by "much larger codebase and team", relative words aren't helpful from a decision tracking standpoint ;). The issues pointed out around cite specific checkpoints, such as a) needing a virtual filing system, b) not having clear ownership of separate teams (would require separate teams), and c) not scaling to hundreds of developers. Should log them clearly here.

File: docs/adr/adr_monorepo.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+*Pro:*
+- The main reason to move to a monorepo set-up is to force us to communicate/coordinate with each other and to feel a shared responsibility for the monorepo. :)
+
+other side-effects
+- deployment is easier since only one script is needed.
+- management of frontend vs. backend versioning is reduced to a minimum.
+
+*Con:*
+Technically, there are problems that come with a Mono-repo set up: slow code search, time for testing, chaotic repo,...
```
Comment (pylipp) [inline docs/adr/adr_monorepo.md:41]: 'chaotic repo' depends on how homogeneous both frontend and backend codebase are to interact with (e.g. in terms of setting up (simply run `./setup.bash --frontend` or `--backend`?) or invoking tests (`./test.bash --frontend/--backend/--all`?). If we'd spend too much effort on aligning these interfaces I'd just keep the polyrepos.

In a monorepo I still see it important to be able to set up and develop frontend and backend code separate from each other. Meaning that anybody who's not doing full-stack development should be able to only set up either part.

**Comment by mcgnly** [comment]
I wouldn't have been in favor of the monorepo until recently, but now that I'm using somethign like that at work it's actually nicer than I thought it would be. I'm totally ok with it now! The main benefit I've seen is that you have one version for a works-together-entire-stack, rather than having to keep track of what version of what works with every other version of the other parts.

File: docs/adr/adr_monorepo.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+Decision Deadline: 10th of July 20
+
+## Status
+
+proposed
+
+## Context or Problem Statement
+
+We are an all remote team with only bi-weekly or monthly team calls. Therefore, communication between the dev team members and transparency for all members is of the highest priority. At the moment, we have the frontend and backend code separated in two repos. We already see that the documentation is not in the same state on both repos and not all the information is in one place. The instructions to set-up the dev environment is distributed and is confusing for new people.
+For a frontend and backend devs coder can work on small tasks completly separate from each other. However, bigger tasks are tougher to coordinate.
```
[the remark this answers is not in the document]
Comment (HaGuesto) [reply docs/adr/adr_monorepo.md:16]: not the best grammer. sorry.

File: docs/adr/adr_monorepo.md (hunk 4 above)
Comment (aerinsol) [reply docs/adr/adr_monorepo.md:41]: Agreed with @pylipp 's second point.

**Comment by aerinsol** [comment]
> Thanks for the write-up!
> 
> I'd like to suggest a hybrid approach: keeping both frontend and backend repos, and introduce a third, top repository that bundles both repositories (e.g. as git submodules).
> Both ends can be developed and tested individually, and the top repo serves for integration testing and 'single-script' deployment. The underlying frontend and backend codebases are automatically versioned since they're pinned in the top repo. Obviously we could version the full stack in the top repo as well.
> 
> I think overall this approach would foster modularity of the codebase.
> Also, what happens if there is another frontend to be introduced (native mobile or desktop) or if we want to build up documentation and tutorial specific to a certain version of boxtribute? These all could live in their individual repos yet they're syncronized by the top repo.
> 
> [This](https://github.com/ActivityWatch/activitywatch) is a project that's been using this approach. I'm sure we could also reach out to the maintainers for about why they structured their codebases this way.

I'm dubious about this. This seems even more complicated to maintain vs. a polyrepo and we want to minimize that as much as possible. Would be interested in hearing @jamescrowley thoughts.

**Comment by cohenlea** [comment]
My previous experience has mostly been with separate back and front repos, which I thought worked fine, but I see valid points being made for both so I wouldn't be opposed to trying the mono-repo idea.

File: docs/adr/adr_monorepo.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (mcgnly) [reply docs/adr/adr_monorepo.md:16]: In a monorepo architecture, frontend and backend developers can work on small tasks completly separately from each other, and coordinate more easily for more complex, far-reaching taske.

File: docs/adr/adr_monorepo.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (mcgnly) [reply docs/adr/adr_monorepo.md:34]: I think that's a bit of a false argument, unless everyone on the team is a full-stack developer and equally capable of dealing with any problem, which isn't likely to ever be the case

Comment (mcgnly) [inline docs/adr/adr_monorepo.md:34]: I would argue that the main benefit of a monorepo is that you have, for each deployed version, a state at which you knew the BE and FE worked together. Maintaining separate versions of each means you also have to maintain a list of compatible versions of each (ie: FE v.2.0 works with BE v.1.3 and up, etc)

File: docs/adr/adr_monorepo.md (hunk 4 above)
Comment (mcgnly) [reply docs/adr/adr_monorepo.md:41]: same, and that's possible. In my work project, we have that- the only place where the 2 really touch is the docker compose file

File: docs/adr/adr_monorepo.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (HaGuesto) [reply docs/adr/adr_monorepo.md:34]: Yes you are right @mcgnly. However, you can say that everybody would be more aware if something is broken. A result of this awareness could be that the developers nudge each other to fix the broken build.

File: docs/adr/adr_monorepo.md [hunk 6]
```diff
@@ -13,7 +13,7 @@ proposed
 ## Context or Problem Statement
 
 We are an all remote team with only bi-weekly or monthly team calls. Therefore, communication between the dev team members and transparency for all members is of the highest priority. At the moment, we have the frontend and backend code separated in two repos. We already see that the documentation is not in the same state on both repos and not all the information is in one place. The instructions to set-up the dev environment is distributed and is confusing for new people.
-For a frontend and backend devs coder can work on small tasks completly separate from each other. However, bigger tasks are tougher to coordinate. 
+In a monorepo architecture, frontend and backend developers can work on small tasks completly separately from each other, and coordinate more easily for more complex, far-reaching task.
```
Comment (aerinsol) [inline docs/adr/adr_monorepo.md:16]: should say "FE and BE developers can **still** work on small tasks completely separate from each other"

File: docs/adr/adr_monorepo.md [hunk 7]
```diff
@@ -31,16 +31,21 @@ For a frontend and backend devs coder can work on small tasks completly separate
...
+- deployment is easier since only one deployment script is needed.
+- management of frontend vs. backend versioning is reduced to a minimum. For each deployed version, you have a state at which you know that FE and BE worked together. 
+- the development set-up simplifies since only one repo needs to be cloned and React could be run from within Docker.
+- all developers are aware when master is broken. If they cannot fix it they could encourage others. 
 
 *Con:*
 Technically, there are problems that come with a Mono-repo set up: slow code search, time for testing, chaotic repo,...
-However, we argue that all the problems really surface with a much larger codebase and team.
+However, we argue that all the problems really surface when 
+- the codebase becomes so large a virtual filing system, such as VFS for Git from Microsoft.
```
Comment (aerinsol) [inline docs/adr/adr_monorepo.md:45]: should say " becomes so large a virtual filing system **is needed**"

File: docs/adr/adr_monorepo.md [hunk 8]
```diff
@@ -31,16 +31,21 @@ For a frontend and backend devs coder can work on small tasks completly separate
...
+- we have hundreds of developers on our team.
```
Comment (aerinsol) [inline docs/adr/adr_monorepo.md:46]: "many more developers on our team, or at least fully separate teams of 7+ people each that have dedicated ownership of specific parts of the repo"  

I imagine we will want to revisit the decision if/when we get to size of 50 people, not 100+ ;)

**Review by aerinsol — COMMENTED** [review]
still awaiting requested changes

**Pull request merged**